network/dag/state.go

```diff
@@ -139,60 +139,88 @@ func (s *state) Add(ctx context.Context, transaction Transaction, payload []byte
-	return s.graph.findBetweenLC(ctx, startInclusive, endExclusive)
+func (s *state) FindBetweenLC(startInclusive uint32, endExclusive uint32) (transactions []Transaction, err error) {
+	err = s.db.View(func(tx *bbolt.Tx) error {
+		transactions, err = s.graph.findBetweenLC(tx, startInclusive, endExclusive)
+		return err
+	})
+	return
 }
 
-func (s *state) GetTransaction(ctx context.Context, hash hash.SHA256Hash) (Transaction, error) {
-	return s.graph.Get(ctx, hash)
+func (s *state) GetTransaction(ctx context.Context, hash hash.SHA256Hash) (transaction Transaction, err error) {
+	err = storage.BBoltTXView(ctx, s.db, func(contextWithTX context.Context, tx *bbolt.Tx) error {
+		transaction, err = s.graph.get(tx, hash)
+		return err
+	})
+	return
 }
 
-func (s *state) IsPayloadPresent(ctx context.Context, hash hash.SHA256Hash) (bool, error) {
-	return s.payloadStore.IsPayloadPresent(ctx, hash)
+func (s *state) IsPayloadPresent(ctx context.Context, hash hash.SHA256Hash) (present bool, err error) {
+	err = storage.BBoltTXView(ctx, s.db, func(contextWithTX context.Context, tx *bbolt.Tx) error {
+		present = s.payloadStore.IsPayloadPresent(tx, hash)
+		return nil
+	})
+	return
 }
 
-func (s *state) IsPresent(ctx context.Context, hash hash.SHA256Hash) (bool, error) {
-	return s.graph.IsPresent(ctx, hash)
+func (s *state) IsPresent(ctx context.Context, hash hash.SHA256Hash) (present bool, err error) {
+	err = storage.BBoltTXView(ctx, s.db, func(contextWithTX context.Context, tx *bbolt.Tx) error {
+		present = s.graph.isPresent(tx, hash)
+		return nil
+	})
+	return
 }
 
-func (s *state) WritePayload(ctx context.Context, transaction Transaction, payloadHash hash.SHA256Hash, data []byte) error {
-	return storage.BBoltTXUpdate(ctx, s.db, func(contextWithTX context.Context, tx *bbolt.Tx) error {
-		err := s.payloadStore.WritePayload(contextWithTX, payloadHash, data)
-		if err == nil {
-			// ctx passed with bbolt transaction
-			return s.notifyPayloadObservers(contextWithTX, transaction, data)
-		}
-		return err
+func (s *state) WritePayload(transaction Transaction, payloadHash hash.SHA256Hash, data []byte) error {
+	return storage.BBoltTXUpdate(context.Background(), s.db, func(contextWithTX context.Context, tx *bbolt.Tx) error {
+		return s.writePayload(tx, transaction, payloadHash, data)
 	})
 }
 
-func (s *state) ReadPayload(ctx context.Context, hash hash.SHA256Hash) ([]byte, error) {
-	return s.payloadStore.ReadPayload(ctx, hash)
+func (s *state) writePayload(tx *bbolt.Tx, transaction Transaction, payloadHash hash.SHA256Hash, data []byte) error {
+	err := s.payloadStore.WritePayload(tx, payloadHash, data)
+	if err == nil {
+		// ctx passed with bbolt transaction
+		return s.notifyPayloadObservers(tx, transaction, data)
+	}
+	return err
+}
+
+func (s *state) ReadPayload(ctx context.Context, hash hash.SHA256Hash) (payload []byte, err error) {
```

Comment:
you changed WritePayload, why not this one as well?

Reply:
As can be seen in the diagram, ReadPayload always starts a transaction.